File: benchmark/java/nla/ProdBin.java
public class ProdBin {
    public static void vtrace1(int a, int b, int x, int y, int z){}
    //public static void vtrace2(int a, int b, int x, int z){}
    public static void main (String[] args) {}
    public static int mainQ(int a, int b){
	assert(a >= 0);	/* SPF/Z3 has some prob without this*/
	assert(b >= 0);

	int x,y,z;
	x = a;
	y = b;
	z = 0;
	
	while(true) { 
	    //assert(z+x*y==a*b);
	    vtrace1(a, b, x, y, z);
	    
	    if(!(y != 0)) break;
	  
	    if (y % 2 == 1){
		z = z+x;
		y = y-1;
	    }
	    x = 2*x;
	    y = y/2;

	}
	
	//assert(z == a*b);
	//do not consider y as it will be 0
	//vtrace2(a, b, x, z);
	return z; 
    }
}

File: benchmark/c/nla/prodbin.c
#include <stdio.h>
#include <stdlib.h>

void vassume(int b){}
void vtrace1(int a, int b, int x, int y, int z){}
void vtrace2(int a, int b, int x, int z){}

int mainQ(int a, int b){
    vassume(a >= 0);
    vassume(b >= 0);

    int x,y,z;
    x = a;
    y = b;
    z = 0;

    while(1) { 
	//assert(z+x*y==a*b);
	//%%%traces: int a, int b, int x, int y, int z
	vtrace1(a, b, x, y, z);
	if(!(y!=0)) break;
	  
	if (y%2 ==1 ){
	    z = z+x;
	    y = y-1;
	}
	x = 2*x;
	y = y/2;

    }
    vtrace2(a, b, x, z);
    //assert(z == a*b);
    return z; 

}


void main(int argc, char **argv){
    mainQ(atoi(argv[1]), atoi(argv[2]));
}

